### PR TITLE
tvk-oneclick

### PR DESCRIPTION
**Pull request by trilio-nikita**
This plugin is specifically for Digital Ocean. It would get added in Digital Ocean marketplace so that on one click, user would be able to install TVK, install free trial license, configure TVK UI, create target and some sample backup & restore tasks.

### CODE REVIEW

File: tests/tvk-oneclick/creat_virtual_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+def create_vcluster()
```
Comment (msurudwad) [inline tests/tvk-oneclick/creat_virtual_cluster.sh:3]: write `cleanup` func as well, which will clean up all the resources and virtual cluster you've created on cluster as a part of CI process.
`cleanup` func should run at the end, irrespective of test job failure/success

File: tests/tvk-oneclick/input_config [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+operator_version= 			#Specify TVK operator version
+triliovault_manager_version='v2.1.1-alpha' #Specify TVK Manager version
+if_resource_exists_still_proceed=True   #Specify if plugin should proceed even if resource exists
+ui_access_type='Loadbalancer' 	#Specify the way in which TVK UI should be configured['Loadbalancer','Nodeport','PortForwarding']
+domain='nsadnanidocean.cf'   	#Specify the domain name, this should be present or added in DNS server
+tvkhost_name='tvkdemo'	      	#Specify TVK host name
+cluster_name='trilio-test2'   	#Specify cluster name
+target_type='NFS'		#Specify Target type ['NFS','DOKs_S3']
+access_key='FJYJK72CCDOD2NB7P5JB'	#Specify access_key
+secret_key='juXtyjY2QS+yKsX4A2pdZ7Sm105EuWmQ1MVmguHZe5c'	#Specify secret key
```
Comment (msurudwad) [inline tests/tvk-oneclick/input_config:17]: this is a public repo... check if this a sensitive info

File: tests/tvk-oneclick/s3cfg_config [hunk 3]
```diff
@@ -0,0 +1,77 @@
+[default]
+access_key = FJYJK72CCDOD2NB7P5JB
```
Comment (msurudwad) [inline tests/tvk-oneclick/s3cfg_config:2]: this is public repo... check if this a sensitive info

File: tests/tvk-oneclick/s3cfg_config [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+proxy_host =
+proxy_port = 0
+put_continue = False
+recursive = False
+recv_chunk = 65536
+reduced_redundancy = False
+requester_pays = False
+restore_days = 1
+restore_priority = Standard
+secret_key = juXtyjY2QS+yKsX4A2pdZ7Sm105EuWmQ1MVmguHZe5c
```
Comment (msurudwad) [inline tests/tvk-oneclick/s3cfg_config:59]: same here

File: tests/tvk-oneclick/tvk_oneclick_test.sh [hunk 5]
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env bash
+
+set -o pipefail
+
+PREFLIGHT_TESTS_SUCCESS=true
```
Comment (msurudwad) [inline tests/tvk-oneclick/tvk_oneclick_test.sh:5]: wrong var name

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: .github/workflows/plugin-packages.yml [hunk 6]
```diff
@@ -256,17 +268,83 @@ jobs:
...
+      - name: Create Virtual cluster
+        run: creat_virtual_cluster.sh
+        with:
+          job_name: "Tvk_oneclclick"
+          install_namespace: "default"
+ 
+       - name: Update 'input_config'
+        run: |
+          echo "Updating 'tests/tvk-oneclick/input_config to run tvk-oneclick"
+          sed -i "s/^\(nfs_server\s*=\s*\).*$/\1\'${{ secrets.NFS_SERVER_IP }}\'/" tests/tvk-oneclick/input_config
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply .github/workflows/plugin-packages.yml:312]: I am updating config file which will be provided as input to a tool. Tool internally will create target.yaml

File: hack/build-tvk-oneclick-artifacts.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+while IFS= read -r -d $'\0' f; do
+  echo "modifying atime/mtime for $f"
+  TZ=UTC touch -at "0001010000" "$f"
+  TZ=UTC touch -mt "0001010000" "$f"
+done < <(find . -print0)
+
+tar -cvzf ${tvk_oneclick_tar_archive} tvk-oneclick/
+echo >&2 "Created ${tvk_oneclick_tar_archive} archive successfully"
+
+# create preflight tar sha256 file
```
Comment (msurudwad) [inline hack/build-tvk-oneclick-artifacts.sh:29]: wrong comment

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: tests/tvk-oneclick/tvk_oneclick_test.sh [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  kubectl delete ns "${INSTALL_NAMESPACE}" --request-timeout 2m || true
+
+  kubectl get po,rs,deployment,pvc,svc,sts,cm,secret,sa,role,rolebinding,job,target,backup,backupplan,policy,restore,cronjob -n "${INSTALL_NAMESPACE}" || true
+
+  kubectl get validatingwebhookconfigurations,mutatingwebhookconfigurations -A | grep -E "${INSTALL_NAMESPACE}" || true
+
+  #Destroying virtual cluster created
+  # shellcheck disable=SC2154
+  vcluster delete "$build_id" -n default
+  helm delete "$build_id" --namespace default
```
Comment (msurudwad) [inline tests/tvk-oneclick/tvk_oneclick_test.sh:142]: wrong sequence of deletion. delete vcluster at the end

File: tests/tvk-oneclick/tvk_oneclick_test.sh [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  rc=$?
+  # shellcheck disable=SC2181
+  if [ $rc != "0" ]; then
+    # shellcheck disable=SC2082
+    echo "Failed - test sample_test, Expected 0 got $rc"
+  fi
+  return $rc
+}
+
+cleanup() {
```
Comment (msurudwad) [inline tests/tvk-oneclick/tvk_oneclick_test.sh:117]: Are you creating these resources that you're trying to delete here?

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: .github/workflows/plugin-packages.yml [hunk 10]
```diff
@@ -293,9 +371,17 @@ jobs:
...
+        uses: actions/download-artifact@v2
+        if: needs.pre-release.outputs.release_tvk_oneclick == 'true'
+        with:
+          name: tvk-oneclick 
+          path: build
+
       - name: Update GoReleaser Config for Release
         run: hack/update-goreleaser-yaml.sh
         env:
+          release_tvk-oneclick: ${{needs.pre-release.outputs.release_tvk_oneclick }}
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply .github/workflows/plugin-packages.yml:384]: Done!

File: tools/tvk-oneclick/tvk-oneclick [hunk 11]
```diff
@@ -0,0 +1,973 @@
...
+#This program is use to install/configure/test TVK product with one click and few required inputs
+
+
+
+#This module is used to perform preflight check which checks if all the pre-requisites are satisfied before installing Triliovault for Kubernetes application in a Kubernetes cluster
+
+preflight_checks()
+{
+  ret=$(kubectl krew 2>/dev/null)
+  if [[ -z "$ret" ]];then
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply tools/tvk-oneclick/tvk-oneclick]: As discussed, we will take this in next release

File: tools/tvk-oneclick/tvk-oneclick [hunk 12]
```diff
@@ -0,0 +1,973 @@
...
+    echo "Please install krew plugin and then try.For information on krew installation please visit:"
+    echo "https://krew.sigs.k8s.io/docs/user-guide/setup/install/"
+    exit 1
+  fi
+  if  [[ $(kubectl tvk-preflight --help  2>/dev/null) ]];
+  then
+    echo "Skipping plugin tvk-preflight installation as it is already installed"
+  else
+    plugin_url='https://github.com/trilioData/tvk-plugins.git'
+    kubectl krew index add tvk-plugins "$plugin_url" >/dev/null 2>/dev/null
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply tools/tvk-oneclick/tvk-oneclick]: done!

File: hack/generate-test-tvk-oneclick-plugin-manifest.sh [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+tvk_oneclick_yaml="tvk-oneclick.yaml"
+cp .krew/$tvk_oneclick_yaml $build_dir/$tvk_oneclick_yaml
+
+tvk_oneclick_yaml=$build_dir/$tvk_oneclick_yaml
+
+tar_checksum="$(awk '{print $1}' $build_dir/tvk-oneclick-sha256.txt)"
+sed -i "s/PREFLIGHT_TAR_CHECKSUM/${tar_checksum}/g" $tvk_oneclick_yaml
+# shellcheck disable=SC2154
+sed -i "s/PREFLIGHT_VERSION/$git_version/g" $tvk_oneclick_yaml
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply hack/generate-test-tvk-oneclick-plugin-manifest.sh:31]: done!

File: .github/workflows/plugin-packages.yml [hunk 14]
```diff
@@ -293,9 +371,17 @@ jobs:
           name: preflight
           path: build
 
+      - name: Download Tvk-oneclick Packages from Artifact
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply .github/workflows/plugin-packages.yml:374]: done!

File: .krew/tvk-oneclick.yaml [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    Curl and Git neeed to be installed on system for Krew index update to work properly
+  platforms:
+  - selector:
+      matchExpressions:
+      - key: os
+        operator: In
+        values:
+        - darwin
+        - linux
+    uri: https://github.com/trilioData/tvk-plugins/releases/download/v1.0.0-rc5/tvk-oneclick.tar.gz
```
[the remark this answers is not in the document]
Comment (trilio-nikita) [reply .krew/tvk-oneclick.yaml:19]: done!

File: hack/build-tvk-oneclick-artifacts.sh (hunk 7 above)
Comment (trilio-nikita) [reply hack/build-tvk-oneclick-artifacts.sh:29]: done!

File: tests/tvk-oneclick/tvk_oneclick_test.sh (hunk 8 above)
Comment (trilio-nikita) [reply tests/tvk-oneclick/tvk_oneclick_test.sh:142]: Corrected!

File: tests/tvk-oneclick/tvk_oneclick_test.sh (hunk 9 above)
Comment (trilio-nikita) [reply tests/tvk-oneclick/tvk_oneclick_test.sh:117]: Yes

File: .github/workflows/plugin-packages.yml [hunk 16]
```diff
@@ -256,17 +268,79 @@ jobs:
...
+          export_default_credentials: true
+
+      - name: Authenticate and Connect to GKE cluster
+        if: (needs.pre-build.outputs.testoneclick == 'true') || startsWith(github.ref, 'refs/tags/')
+        uses: google-github-actions/get-gke-credentials@main
+        with:
+          cluster_name: ${{ env.GKE_CLUSTER }}
+          location: ${{ env.GKE_ZONE }}
+          use_auth_provider: true
+          credentials: ${{ secrets.GKE_SA_KEY }}
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:301]: Comment this two

File: .github/workflows/plugin-packages.yml [hunk 17]
```diff
@@ -256,17 +268,79 @@ jobs:
...
+
+      - name: Update 'input_config'
+        run: |
+          echo "Updating 'tests/tvk-oneclick/input_config to run tvk-oneclick"
+          sed -i "s/^\(nfs_path\s*=\s*\).*$/\1\'\/src\/nfs\/kubedata\'/" tests/tvk-oneclick/input_config
+          nfs_server=${{ secrets.NFS_SERVER_IP }}
+          sed -i "s/^\(nfs_server\s*=\s*\).*$/\1\'$nfs_server\'/" tests/tvk-oneclick/input_config
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:308]: it's used in test only right.. comment this, if yes

File: .github/workflows/plugin-packages.yml [hunk 18]
```diff
@@ -286,16 +360,17 @@ jobs:
...
 
-      - name: Download Preflight Packages from Artifact
+      - name: Download Preflight/Tvk-oneclick Packages from Artifact
         uses: actions/download-artifact@v2
-        if: needs.pre-release.outputs.release_preflight == 'true'
+        if: (needs.pre-release.outputs.release_preflight == 'true') || (needs.pre-release.outputs.release_tvk_oneclick == 'true')
         with:
-          name: preflight
+          name: plugin packages
           path: build
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:368]: this is wrong.. refer doc of `actions/download-artifact@v2`

File: docs/tvk-oneclick/README.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+**Supported OS and Architectures**:
+
+OS:
+- Linux
+- darwin
+
+Arch:
+- amd64
+- x86
```
Comment (msurudwad) [inline docs/tvk-oneclick/README.md:25]: no need of this.. remove

File: hack/update-goreleaser-yaml.sh [hunk 20]
```diff
@@ -29,5 +31,10 @@ if [[ $release_target_browser != true ]]; then
   sed -i '/binary: target-browser/a \ \ skip: true' "$goreleaser_yaml"
 fi
 
+if [[ $release_tvk_oneclick != true ]]; then
+  echo "skip tvk-oneclick packages release from goreleaser.yml"
+  sed -i '/binary: tvk-oneclick/a \ \ skip: true' "$goreleaser_yaml"
+fi
```
Comment (msurudwad) [inline hack/update-goreleaser-yaml.sh:37]: this is wrong

File: .github/workflows/plugin-packages.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (msurudwad) [reply .github/workflows/plugin-packages.yml:373]: also, `hack/check-git-diff-between-tags.sh` needs updates for your plugin

File: .github/workflows/plugin-packages.yml [hunk 21]
```diff
@@ -286,16 +360,17 @@ jobs:
...
         uses: actions/checkout@v2
 
-      - name: Download Preflight Packages from Artifact
+      - name: Download Preflight/Tvk-oneclick Packages from Artifact
         uses: actions/download-artifact@v2
-        if: needs.pre-release.outputs.release_preflight == 'true'
+        if: (needs.pre-release.outputs.release_preflight == 'true') || (needs.pre-release.outputs.release_tvk_oneclick == 'true')
         with:
-          name: preflight
+          name: plugin packages
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:367]: There no artifact as such.. from did you get this name `plugin packages`

**Review by msurudwad — CHANGES_REQUESTED** [review]

**Review by msurudwad — APPROVED** [review]

**Pull request merged**